QccPack-0.61-1/src/libQccPackIMG/utils/man1/seqdist.1
.TH SEQDIST 1 "QCCPACK" ""
.SH NAME
seqdist \- calculate the distortion between two image sequences
.SH SYNOPSIS
.B seqdist
.RB "[\|" \-vo "\|]"
.RB "[\|" \-sf
.IR  startframe "\|]"
.RB "[\|" \-ef
.IR  endframe "\|]"
.RB "[\|" \-mse "\|]"
.RB "[\|" \-snr "\|]"
.RB "[\|" \-psnr "\|]"
.RB "[\|" \-mae "\|]"
.RB "[\|" \-w "\|]"
.I seqfile1
.I seqfile2
.SH OPTIONS
.TP
.B \-vo
Output only the value of the distortion (terse output).
.TP
.B \-sf
Start frame for subsequence over which to calculate distortion.
Default is first available frame.
.TP
.B \-ef
End frame for subsequence over which to calculate distortion.
Default is last available frame.
.TP
.B \-mse
Express the distortion in terms of MSE.
.TP
.B \-snr
Express the distortion in terms of SNR.
.TP
.B \-psnr
Express the distortion in terms of PSNR.
.TP
.B \-mae
Express the distortion in terms of MAE.
.TP
.B \-w
Output the distortion over all frames of the sequence only.
.SH DESCRIPTION
.LP
.B seqdist
prints
the mean square error (MSE), signal-to-noise ratio (SNR),
peak signal-to-noise ratio (PSNR), or
maximum absolute error (MAE)
between image sequences
.I seqfile1
and
.I seqfile2
to stdout.
.LP
The options
.BR \-mse ,
.BR \-snr ,
.BR \-psnr ,
and
.BR \-mae 
indicate in which form the distortion should be expressed.
If none of these
options are specified, 
.B seqdist
prints the distortion in all forms.
.LP
If the option
.B \-vo
is specified,
then 
.B seqdist
prints a terse output, consisting of only the value of the distortion,
to stdout.  One of the options
.BR \-mse ,
.BR \-snr ,
.BR \-psnr ,
or
.BR \-mae
must be specified in this case.
.LP
The options
.B \-sf
and
.B \-ef
specify the starting and ending frame numbers, respectively, 
over which to calculate the distortion. If either of these
is not specified,
.BR seqdist
determines the first and/or last frame automatically from the files that
exist via
.BR QccIMGImageSequenceFindFrameNums (3).
.LP
If the 
.B \-w
option is not given, 
the distortion over each individual frame is printed.  Additionally,
in the case of verbose output 
.RB ( \-vo
not given), the distortion over the entire sequence (or relevant
subsequence, in the case that
.B \-sf
or
.B \-ef
are specified) is output.
If
.B \-w
is given, only this latter distortion is output.
.LP
If the image files are color,
the distortion calculations are performed separately for the Y, U, and V 
components of the image.  The distortions for each image component are
output in order.
.LP
For the calculation of SNR and PSNR,
.B seqfile1
is considered the original signal; 
.B seqdist
thus finds the variance of
.B seqfile1
for use in the SNR calculation.
For the calculation of PSNR,
.B seqdist
assumes that the peak-to-peak range of
.B seqfile1
is 255 levels.
.SH "SEE ALSO"
.BR imgdist (1),
.BR QccIMGImageComponentMSE (3),
.BR QccIMGImageComponentMAE (3),
.BR QccIMGImageSequence (3),
.BR QccPackIMG (3),
.BR QccPack (3)

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackIMG/utils/seqdist.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "seqdist.h"

#define USG_STRING "[-vo %:] [-sf %: %d:startframe] [-ef %: %d:endframe] [-mse %:] [-snr %:] [-psnr %:] [-mae %:] [-w %:] %s:seqfile1 %s:seqfile2"

#define ALL 0
#define MSE 1
#define SNR 2
#define PSNR 3
#define MAE 4

int StartFrameNumSpecified = 0;
int StartFrameNum;
int EndFrameNumSpecified = 0;
int EndFrameNum;

int value_only_flag = 0;
int mse_specified = 0;
int snr_specified = 0;
int psnr_specified = 0;
int mae_specified = 0;

int whole_sequence_only = 0;

int DistortionMethod = ALL;

int ImageType;
int MaxVal = 255;

int NumCols1, NumCols2;
int NumRows1, NumRows2;

int NumFrames;

QccIMGImageSequence ImageSequence1;
QccIMGImageSequence ImageSequence2;


int main(int argc, char *argv[])
{
  double value = 0;
  double mse = -1.0;
  double snr = -1.0;
  double psnr = -1.0;
  double mae = -1.0;
  double signal_power;
  QccIMGImageComponent *image_component1 = NULL;
  QccIMGImageComponent *image_component2 = NULL;
  int component, num_components;
  QccString component_name;
  int num_rows1_Y, num_cols1_Y;
  int num_rows1_U, num_cols1_U;
  int num_rows1_V, num_cols1_V;
  int num_rows2_Y, num_cols2_Y;
  int num_rows2_U, num_cols2_U;
  int num_rows2_V, num_cols2_V;
  int current_frame;
  double total_mse[3] = { 0.0, 0.0, 0.0 };
  double total_mae[3] = { 0.0, 0.0, 0.0 };
  double total_signal_power[3] = { 0.0, 0.0, 0.0 };
  double total_snr = 0.0;
  double total_psnr = 0.0;

  QccInit(argc, argv);

  QccIMGImageSequenceInitialize(&ImageSequence1);
  QccIMGImageSequenceInitialize(&ImageSequence2);
  
  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         &value_only_flag,
                         &StartFrameNumSpecified,
			 &StartFrameNum,
                         &EndFrameNumSpecified,
			 &EndFrameNum,
                         &mse_specified,
                         &snr_specified,
                         &psnr_specified,
                         &mae_specified,
                         &whole_sequence_only,
                         ImageSequence1.filename, ImageSequence2.filename))
    QccErrorExit();

  if (mse_specified + snr_specified + psnr_specified + mae_specified > 1)
    {
      QccErrorAddMessage("%s: Only one of -mse, -snr, -psnr, and -mae can be given",
                         argv[0]);
      QccErrorExit();
    }

  if (mse_specified)
    DistortionMethod = MSE;
  else if (snr_specified)
    DistortionMethod = SNR;
  else if (psnr_specified)
    DistortionMethod = PSNR;
  else if (mae_specified)
    DistortionMethod = MAE;
  else
    DistortionMethod = ALL;

  if (value_only_flag &&
      (!mse_specified && !snr_specified && !psnr_specified && !mae_specified))
    {
      QccErrorAddMessage("%s: One of -mse, -snr, -psnr, and -mae must be used with -vo",
                         argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageSequenceFindFrameNums(&ImageSequence1))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageSequenceFindFrameNums()",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageSequenceFindFrameNums(&ImageSequence2))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageSequenceFindFrameNums()",
                         argv[0]);
      QccErrorExit();
    }

  if (!StartFrameNumSpecified)
    StartFrameNum = ImageSequence1.start_frame_num;

  if ((StartFrameNum < ImageSequence1.start_frame_num) ||
      (StartFrameNum < ImageSequence2.start_frame_num))
    {
      QccErrorAddMessage("%s: Start frame %d is out of the range of available frames which start at %d",
                         argv[0],
                         StartFrameNum,
                         QccMathMax(ImageSequence1.start_frame_num,
                                    ImageSequence2.start_frame_num));
      QccErrorExit();
    }
  ImageSequence1.start_frame_num = StartFrameNum;
  ImageSequence2.start_frame_num = StartFrameNum;

  if (!EndFrameNumSpecified)
    EndFrameNum = ImageSequence1.end_frame_num;

  if ((EndFrameNum > ImageSequence1.end_frame_num) ||
      (EndFrameNum > ImageSequence2.end_frame_num))
    {
      QccErrorAddMessage("%s: End frame %d is out of the range of available frames which end at %d",
                         argv[0],
                         EndFrameNum,
                         QccMathMin(ImageSequence1.end_frame_num,
                                    ImageSequence2.end_frame_num));
      QccErrorExit();
    }
  ImageSequence1.end_frame_num = EndFrameNum;
  ImageSequence2.end_frame_num = EndFrameNum;

  NumFrames = QccIMGImageSequenceLength(&ImageSequence1);
  if (NumFrames !=
      QccIMGImageSequenceLength(&ImageSequence2))
    {
      QccErrorAddMessage("%s: Image sequences are of different lengths",
                         argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageSequenceStartRead(&ImageSequence1))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageSequenceStartRead()",
                         argv[0]);
      QccErrorExit();
    }
  ImageType = ImageSequence1.current_frame.image_type;

  if (QccIMGImageSequenceStartRead(&ImageSequence2))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageSequenceStartRead()",
                         argv[0]);
      QccErrorExit();
    }
  if (ImageSequence2.current_frame.image_type != ImageType)
    {
      QccErrorAddMessage("%s: Image sequences must both be color (or grayscale)",
                         argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageGetSizeYUV(&(ImageSequence1.current_frame),
                            &num_rows1_Y, &num_cols1_Y,
                            &num_rows1_U, &num_cols1_U,
                            &num_rows1_V, &num_cols1_V))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSizeYUV()",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageGetSizeYUV(&(ImageSequence2.current_frame),
                            &num_rows2_Y, &num_cols2_Y,
                            &num_rows2_U, &num_cols2_U,
                            &num_rows2_V, &num_cols2_V))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSizeYUV()",
                         argv[0]);
      QccErrorExit();
    }

  num_components = QccIMGImageColor(&(ImageSequence1.current_frame)) ? 3 : 1;

  if (!value_only_flag)
    printf("Distortion between sequences\n   %s\n         and\n   %s\nis:\n",
           ImageSequence1.filename, ImageSequence2.filename);

  
  ImageSequence1.current_frame_num = ImageSequence1.start_frame_num - 1;
  ImageSequence2.current_frame_num = ImageSequence2.start_frame_num - 1;

  for (current_frame = ImageSequence1.start_frame_num;
       current_frame <= ImageSequence1.end_frame_num;
       current_frame++)
    {
      if (QccIMGImageSequenceIncrementFrameNum(&ImageSequence1))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageSequenceIncrementFrameNum()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageSequenceReadFrame(&ImageSequence1))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageSequenceReadFrame()",
                             argv[0]);
          QccErrorExit();
        }

      if (QccIMGImageSequenceIncrementFrameNum(&ImageSequence2))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageSequenceIncrementFrameNum()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageSequenceReadFrame(&ImageSequence2))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageSequenceReadFrame()",
                             argv[0]);
          QccErrorExit();
        }

      for (component = 0; component < num_components; component++)
        {
          switch (component)
            {
            case 0:
              image_component1 = &(ImageSequence1.current_frame.Y);
              image_component2 = &(ImageSequence2.current_frame.Y);
              QccStringCopy(component_name, "Y component");
              NumRows1 = num_rows1_Y;
              NumCols1 = num_cols1_Y;
              NumRows2 = num_rows2_Y;
              NumCols2 = num_cols2_Y;
              break;
            case 1:
              image_component1 = &(ImageSequence1.current_frame.U);
              image_component2 = &(ImageSequence2.current_frame.U);
              QccStringCopy(component_name, "U component");
              NumRows1 = num_rows1_U;
              NumCols1 = num_cols1_U;
              NumRows2 = num_rows2_U;
              NumCols2 = num_cols2_U;
              break;
            case 2:
              image_component1 = &(ImageSequence1.current_frame.V);
              image_component2 = &(ImageSequence2.current_frame.V);
              QccStringCopy(component_name, "V component");
              NumRows1 = num_rows1_V;
              NumCols1 = num_cols1_V;
              NumRows2 = num_rows2_V;
              NumCols2 = num_cols2_V;
              break;
            }          
          
          if ((NumCols1 != NumCols2) || (NumRows1 != NumRows2))
            {
              QccErrorAddMessage("%s: Size of %s (%d x %d) does not equal that of %s (%d x %d) for image component %d",
                                 argv[0],
                                 ImageSequence1.filename, NumCols1, NumRows1,
                                 ImageSequence2.filename, NumCols2, NumRows2,
                                 component);
              QccErrorExit();
            }
          
          mse = QccIMGImageComponentMSE(image_component1, image_component2);
          mae = QccIMGImageComponentMAE(image_component1, image_component2);
          total_mse[component] += mse/NumFrames;
          total_mae[component] += mae/NumFrames;
          signal_power = QccIMGImageComponentVariance(image_component1);
          total_signal_power[component] += signal_power/NumFrames;

          if (mse != 0.0)
            {
              psnr = 10.0*log10((double)MaxVal*MaxVal/mse);
              snr = 10.0*log10(signal_power/mse);
            }
          
          if (!whole_sequence_only)
            {
              if (!value_only_flag)
                {
                  printf("  Frame #%d:\n", current_frame);
                  if (num_components != 1)
                    printf("   %s:\n", component_name);
                  switch (DistortionMethod)
                    {
                    case MSE:
                      printf("%16.6f (MSE)\n", mse);
                      break;
                    case SNR:
                      printf("%16.6f dB (SNR)\n", snr);
                      break;
                    case PSNR:
                      printf("%16.6f dB (PSNR)\n", psnr);
                      break;
                    case MAE:
                      printf("%16.6f (MAE)\n", mae);
                      break;
                    case ALL:
                    default:
                      printf("%16.6f\t(MSE)\n%16.6f dB\t(SNR)\n%16.6f dB\t(PSNR)\n%16.6f\t(MAE)\n",
                             mse, snr, psnr, mae);
                      break;
                    }
                }
              else
                {
                  switch (DistortionMethod)
                    {
                    case MSE:
                      value = mse;
                      break;
                    case SNR:
                      value = snr;
                      break;
                    case PSNR:
                      value = psnr;
                      break;
                    case MAE:
                      value = mae;
                      break;
                    }
                  printf("%f\n", value);
                }
            }
        }
    }

  if (whole_sequence_only || !value_only_flag)
    {
      if (!value_only_flag)
        {
          if (!whole_sequence_only)
            printf("===========================================================\n");
          printf("Over whole sequence:\n");
        }

      for (component = 0; component < num_components; component++)
        {
          if (total_mse[component] != 0.0)
            {
              total_psnr = 
                10.0*log10((double)MaxVal*MaxVal/total_mse[component]);
              total_snr = 
                10.0*log10(total_signal_power[component]/total_mse[component]);
            }
          if (!value_only_flag)
            {
              if (num_components != 1)
                printf("   %s:\n", component_name);
              switch (DistortionMethod)
                {
                case MSE:
                  printf("%16.6f (MSE)\n", total_mse[component]);
                  break;
                case SNR:
                  printf("%16.6f dB (SNR)\n", total_snr);
                  break;
                case PSNR:
                  printf("%16.6f dB (PSNR)\n", total_psnr);
                  break;
                case MAE:
                  printf("%16.6f (MAE)\n", total_mae[component]);
                  break;
                case ALL:
                default:
                  printf("%16.6f\t(MSE)\n%16.6f dB\t(SNR)\n%16.6f dB\t(PSNR)\n%16.6f\t(MAE)\n",
                         total_mse[component], total_snr, total_psnr,
                         total_mae[component]);
                  break;
                }
            }
          else
            {
              switch (DistortionMethod)
                {
                case MSE:
                  value = total_mse[component];
                  break;
                case SNR:
                  value = total_snr;
                  break;
                case PSNR:
                  value = total_psnr;
                  break;
                case MAE:
                  value = total_mae[component];
                  break;
                }
              printf("%f\n", value);
            }
        }
    }

  QccIMGImageSequenceFree(&ImageSequence1);
  QccIMGImageSequenceFree(&ImageSequence1);

  QccExit;
}
